Doc/library/functools.rst

```diff
@@ -512,6 +518,167 @@ The :mod:`functools` module defines the following functions:
    .. versionadded:: 3.8
 
 
+.. class:: TopologicalSorter(graph=None)
+
+   Provides functionality to topologically sort a graph of hashable nodes.
+
+   A topological order is a linear ordering of the vertices in a graph such that for
+   every directed edge u -> v from vertex u to vertex v, vertex u comes before vertex
+   v in the ordering. For instance, the vertices of the graph may represent tasks to
+   be performed, and the edges may represent constraints that one task must be
+   performed before another; in this example, a topological ordering is just a valid
+   sequence for the tasks. A complete topological ordering is possible if and only if
+   the graph has no directed cycles, that is, if it is a directed acyclic graph.
+
+   If the optional *graph* argument is provided it must be a dictionary representing
+   a direct acyclic graph where the keys are nodes and the values are iterables of
+   all predecessors of that node in the graph (the nodes that have edges that point
+   to the value in the key). Additional nodes can be added to the graph using the
+   :meth:`~TopologicalSorter.add` method.
+
+   In the general case, the steps required to perform the sorting of a given graph
+   are as follows:
+
+         * Create an instance of the :class:`TopologicalSorter` with an optional
+            initial graph.
+         * Add additional nodes to the graph.
+         * Call "finalize()" on the graph.
```

Comment:
```suggestion
         * Call "prepare()" on the graph.
```

I'm not invested in the "prepare" name, but since it's called that instead of "finalize" everywhere else for now, just keeping it in synch.